### ISSUE TITLE
URL handlers for ckan://

### ISSUE THREAD

**Issue by AlexanderDzhoganov** [KSP-CKAN/CKAN#723]
https://github.com/KSP-CKAN/CKAN-GUI/pull/19 adds handlers for ckan://<identifier> for Windows and Linux. Not sure if this will work on OSX, maybe we need a separate config for it. The handler itself will run `ckan.exe gui %u` where ckan.exe is the location of the most recently used executable and %u is the identifier part.

**Comment by AlexanderDzhoganov** [comment]
Using [this release](https://github.com/KSP-CKAN/CKAN-nightly/releases/tag/jenkins-CKAN-158) you'll be able to click [here](ckan://FerramAerospaceResearch) and have CKAN open.

**Comment by pjf** [comment]
Marking this as closed, because we both have ckan:// handlers, and because they either work or fail in a non-user-provoking way. :)

**Comment by Dazpoet** [comment]
From what I gather these where added so that KerbalStuff (or any other modwebsite for that part, KerbalTopia and KerbalX come to mind) would be able to have a button which led directly to CKAN client starting and marking the mod for installation. Do these handlers acctually work on all OS now or is it an unfinished feature? The question about these are raised from time to time on irc by @SirCmpwn

**Comment by SirCmpwn** [comment]
Yes, they weren't usable last time anyone bothered to answer my questions about them.

**Comment by Dazpoet** [comment]
Reopening so we can evalute the scope of this before opening a new issue with more relevant information. I'm not sure who'd know more save for @AlexanderDzhoganov though.

I assume the initial scope was to add the ability to click something like "Install mod through CKAN" on KerbalStuff which might help KS with https://github.com/KerbalStuff/KerbalStuff/issues/5 and give us a more obvious connection to where users would go looking for usage stats such as which mods are the most popular right now.

**Comment by pjf** [comment]
I've not seen them work on my (Linux) machine, the ini parser library is unable to parse my `mimeapps.list` file. On other platforms I can't comment as to whether they work or not.

**Comment by Halbann** [comment]
A short while later...

I'm interested in implementing this for a dependency helper within Lazy Painter. Some design questions:

- Is this actually something anyone still wants besides myself?
- Would simply `ckan://identifier/install` be an appropriate design for this? It's tempting to find a way to give more control (in a way that's backwards compatible with `ckan://identifier`), but maybe that's not desired.
- How much of the install process is it appropriate to do automatically? Could range from just selecting the mod for the user, to selecting and taking them to the changeset screen, to going through with the installation automatically. I want users to sign off on what I'm installing and be in control if anything goes wrong, so I'd prefer the middle option.
- Opening a ckan URL while CKAN is running should communicate with the running process. Are pipes the way you do this? If so I'll learn how to use them.
- Does this feature need tests? If so, what tests?

**Comment by Halbann** [comment]
Very naively written initial implementation.

https://github.com/user-attachments/assets/c8cde70b-fcb2-470a-bcfa-2b1f2a524d74

**Comment by HebaruSan** [comment]
> * Is this actually something anyone still wants besides myself?

I have never heard of anyone using this feature, so you might be the first. 🥇

> * Would simply `ckan://identifier/install` be an appropriate design for this? It's tempting to find a way to give more control (in a way that's backwards compatible with `ckan://identifier`), but maybe that's not desired.

We could probably get away with breaking backwards compatibility by changing what `ckan://identifier` does, if the old behavior was useless and unused and the new behavior provided more value.

> * How much of the install process is it appropriate to do automatically? Could range from just selecting the mod for the user, to selecting and taking them to the changeset screen, to going through with the installation automatically. I want users to sign off on what I'm installing and be in control if anything goes wrong, so I'd prefer the middle option.

I agree with jumping to the changeset, so the user can see what the changes would be and choose whether to proceed.

A further idea: What would you think of some mechanism for encoding _multiple_ identifiers into the URL?
- `ckan://mod1&mod2&mod3&mod4` (`&` characters probably aren't allowed in that part of the URL, so this is just an example rather than a final proposal)

That way we could potentially simplify certain small- and medium-sized modpacks down to one clickable link. Possibly including on SpaceDock.

> * Opening a ckan URL while CKAN is running should communicate with the running process. Are pipes the way you do this? If so I'll learn how to use them.

I thought that was already how it worked, so I would have to investigate to answer this.

> * Does this feature need tests? If so, what tests?

Everything should be tested unless it's impossible. The easiest path is probably to just make a PR, wait for Coveralls to post a reply highlighting the uncovered stuff, and sort out what to do at that point.

**Comment by HebaruSan** [comment]
> * Opening a ckan URL while CKAN is running should communicate with the running process. Are pipes the way you do this? If so I'll learn how to use them.

The tricky thing here is making it work on all platforms. Or at least Windows and Linux; Mac is kind of its own separate beast these days, since there's no GUI there (and still a `//TODO` in `URLHandlers.cs`), but if your mod's links became popular, it's safe to assume a Mac user would click it someday and expect it to work.

Off the top of my head, I don't think "pipes" would be the solution I would choose, but maybe there's a great way of doing it of which I'm not aware.

**Comment by Halbann** [comment]
Thanks, all great answers. Stringing multiple identifiers would be really nice. From what I can gather, I'm fairly confident that the inter-process communication can be made cross-platform, although I haven't yet looked at how the URL handlers are registered on each platform. I can test it out on Linux tomorrow.

Relating to the URL handler itself, on Windows it simply runs `ckan.exe gui %1`, which creates another instance that inevitably runs into the lock file. In the naive implementation, I'm using the pipe itself to check (only in the case that the args match `gui URL`) if another CKAN is already listening, in which case we stream the URL to them and close immediately.

One thing I don't love about that, at least as it appears on Windows, is the momentary appearance of the console window. I think I could solve that with an additional executable called something like `ckanhandler` (as in `nxmhandler` from nexus mod manager, which probably uses the same trick) with `OutputType` set to `WinExe`. I think it's really nice that CKAN is one self-contained executable though, at least when you download it, so I'm not sure what to do there.

**Comment by HebaruSan** [comment]
One thing to consider is that the link should probably still work if CKAN is **not** already running. I don't think that would work with the `ckanhandler` idea, since it would need to have the ability to launch CKAN, which means it would need to be able to _find_ `ckan.exe` on disk, which is probably not achievable in the great variety of weird ways in which users use CKAN.

**Comment by Halbann** [comment]
Got it working on Linux this evening, so two down and one to go.

https://github.com/user-attachments/assets/58bece33-5e2b-4340-91b0-52e178d1171b

I've pushed the [changes](https://github.com/Halbann/CKAN/compare/fix/url-handler...url-handler-improvements) on a branch of my fork. I want to stress that this is just a fumbling in the dark prototype from someone who's really only used to writing C# for KSP/Unity. I don't have much idea about how things should be and want to redo it in a way that's in keeping with the rest of the code base. If you have any thoughts about how that should look then they'd be much appreciated.

Can't work on this for the next two weeks but this is what I'd like to do when I get back (although any help is welcome):

- Finish #4603, including handling for the console UI.
- Disentangle the single instance launch and the URL handoff from GUI. Make GUI and console UI users of the same system.
- Agree on a scheme for the URL, I was thinking something like this:
  - ckan://[operation]/[argument]
  - operation: focus | search | select | changeset | install
  - argument: `[identifier]` or `[identifier1]~[identifier2]~[etc.]` or `[searchterm]`.
  - I can't think of great use cases for `install` or `select` to be honest. Maybe those don't need to be there.
  - I think search would be really useful (i.e. "here are all \[my mods](ckan://search/%40Halban) on CKAN"), but a URI encoded search term can get quite hard to read and write.
- Investigate methods of handling the URL without creating a console window on Windows.
  - `ckanhandler.exe` that is extracted from `ckan.exe` on launch on Windows whenever `ckanhandler.exe` is missing from the current directory. But maybe that's too voodoo...
  - Continue investigating methods of hiding the console from inside `ckan.exe`.
  - Change the handler command in some way such that `ckan.exe` is launched without creating a window.
- Handle the case where the user opens many ckan URLs in quick succession (spams a button).
- Handle the case where the the user opens a second UI/GUI instance manually. Who owns the pipe now?
- Make `ckan://changeset/` add to the current changeset (and recompute) so that the user can accumulate mods by clicking links.
- Ask the OS to focus the console UI/GUI when receiving a URL.
- Implement the rest of the `ckan://` scheme.
- Write tests. Might be complicated to test the inter-process communication?

**Comment by Halbann** [comment]
> but a URI encoded search term can get quite hard to read and write

Perhaps we could add a button in the search drop down that copies the current search term to the clipboard as `ckan://search/[searchterm]` with the URI encoding done for you.

**Comment by Halbann** [comment]
Updated proposal:
```url
ckan://focus?mod=identifier
ckan://search?q=searchterm
ckan://changeset?mod=id1&mod=id2
ckan://changeset?mod=id1:version&mod=id2
```

Turns out there's a standard way of doing this that I was totally unaware of. This greatly simplifies the code because we can rely on `Uri.TryCreate` and `HttpUtility.ParseQueryString` to do the parsing, and then it's just a matter of invoking an event for each operation or `Host` in URI terms. Version pinning is supported via colon syntax (`mod=identifier:1.2`). Omitting the version is equivalent to requesting the latest.

**Comment by HebaruSan** [comment]
Could we use `install` instead of `changeset`? I feel like it's better to specify _what_ rather than _how_, in case we need to change some of this in the future.

**Comment by Halbann** [comment]
My reasoning with that is that we may wish to add `install` later, meaning install with no user approval, as opposed to bringing the user to the changeset screen with the specified mods. I think that's preferrable because the _what_ is more specific in that case. Although If we (you) ruled out ever including an `install` operation then I'd prefer `install` to mean what I mean by `changeset`.

**Comment by HebaruSan** [comment]
I don't think we can skip the changeset, because it is important for showing the user conflicts.

**Comment by Halbann** [comment]
That's true, and on second thought it's probably not a good idea, if we were ever going to do a zero-oversight install operation, to not make it explicit in the name. Wouldn't want people using it unintentionally.
